Add `quote` element

tingerrr opened the pull request:
This PR adds a quote element and resolves #105.

The quote element has the following parameters:
- `block:  bool           = false`, block or inline
- `quotes: bool | auto    = auto`, add quotes or not, if `auto` then `!block`
- `source: none | str     = none`, the destination URL to wrap this quote in
- `author: none | content = none`, the quote author
- `body: content`

An inline quote is formatted like so and respects text direction, author and quote are swapped on `text(dir: rtl)`:
```typst
And I quote: #quote(author: [René Descartes])[cogito, ergo sum], because it is a great quote.
```
![image](https://github.com/typst/typst/assets/137803093/ac10844d-df9b-40d1-b976-0e2a7b382866)

A block quote is formatted similar to quotes commonly used in emails:
```typst
#quote(author: [René Descartes], block: true)[cogito, ergo sum]
```
![image](https://github.com/typst/typst/assets/137803093/2f67379d-473c-4213-94d9-e8068364b7ed)

As you can see, `quotes: auto` does not yet work as expected when passing `block: true` explicitly. I suspect I need to resolve to `Smart<bool>` and check in `Show::show` whether this was `Smart::Auto` or something similar. I looked at `TextDir` for inspiration here.

I'm also not entirely sure if the spacing of the block element and the author when given is satisfactory.

The `source` parameter is motivated by the [html blockquote](https://developer.mozilla.org/docs/Web/HTML/Element/blockquote), as seen in #105.

## Future Work
Ideally, after work has finished on this, it could be used as basis for introducing dedicated syntax at least for block quotes using:
```
> Blockquote
> > Nested Blockquote
> Blockquote
> Blockquote
```

tingerrr commented:
The clippy failure in CI is about collapsing nested if blocks, but I think that a) the code gets less readable and b) will likely change with the next commit to make clippy happier.

tingerrr commented:
The way `Finalize` is currently used is also definitely not correct as it would interfere with user-defined show rules. That was more of an attempt to make the stroke color settable `show quote: set block(stroke: (left: red))`.

tingerrr commented:
Block quotes now look like this by default:
![image](https://github.com/typst/typst/assets/137803093/89c4fe36-a46d-4590-8248-176eddab5ac1)

And allow setting the alignment of the body `show quote: set align(center)` without affecting the author alignment being `end`:  
![image](https://github.com/typst/typst/assets/137803093/316478c4-78d1-4526-a523-f465619c927d)

laurmaedje commented on crates/typst-library/src/text/quote.rs:
```diff
@@ -0,0 +1,136 @@
+use super::{SmartquoteElem, SpaceElem, TextElem};
+use crate::{
+    layout::{BlockElem, HElem, PadElem, Spacing},
+    prelude::*,
+};
```
could you unnested the imports? we generally don't use the nested imports in the code base.

laurmaedje commented on crates/typst-library/src/text/quote.rs:
```diff
@@ -0,0 +1,136 @@
+    /// - `{true}`: Wrap this quote in double quotes.
+    /// - `{false}`: Do not wrap this quote in double quotes.
+    /// - `{auto}`: Infer whether to wrap this quote in double quotes based on
+    ///   the `block` property. If `block` is `{false}` double quotes are
```
```suggestion
    ///   the `block` property. If `block` is `{false}`, double quotes are
```

laurmaedje commented on crates/typst-library/src/text/quote.rs:
```diff
@@ -0,0 +1,136 @@
+    /// #set smartquote(quotes: "«»")
+    /// #set quote(quotes: true)
+    /// #quote(block: true)[
+    ///   ... ἔοικα γοῦν τούτου γε σμικρῷ τινι αὐτῷ τούτῳ σοφώτερος εἶναι, ὅτι
+    ///   ἃ μὴ οἶδα οὐδὲ οἴομαι εἰδέναι.
```
lang is still "de". that doesn't seem ideal.

tingerrr replied:
Yeah, this was more so to show that smartquote.quotes are used, I'll use a german quote instead.

laurmaedje commented on crates/typst-library/src/text/quote.rs:
````diff
@@ -0,0 +1,136 @@
+    /// ```
+    quotes: Smart<bool>,
+
+    /// The source url to this quote.
````
```suggestion
    /// The source URL to this quote.
```

laurmaedje commented on crates/typst-library/src/text/quote.rs:
````diff
@@ -0,0 +1,136 @@
+    ///
+    /// ```example
+    /// #show link: set text(blue)
+    /// #quote(source: "https://google.com")[cogito, ergo sum]
````
how is google the source for that?

tingerrr replied:
Right, that was residue from testing.

laurmaedje commented on crates/typst-library/src/text/quote.rs:
````diff
@@ -0,0 +1,136 @@
+    /// #show link: set text(blue)
+    /// #quote(source: "https://google.com")[cogito, ergo sum]
+    /// ```
+    source: Option<EcoString>,
````
should this really force a URL? what about bibliography references?

tingerrr replied:
Should I simply check for the string to match `https?://` and else assume it's a bib key then?

laurmaedje commented on crates/typst-library/src/text/quote.rs:
```diff
@@ -0,0 +1,136 @@
+        let author = self.author(styles);
+        let block = self.block(styles);
+
+        if self.quotes(styles).as_custom().is_some_and(|q| q) || !block {
```
```suggestion
        if self.quotes(styles) == Smart::Custom(true) || !block {
```

laurmaedje commented on crates/typst-library/src/text/quote.rs:
```diff
@@ -0,0 +1,136 @@
+                .aligned(Align::END);
+            }
+
+            let pad: Rel = Em::new(1.0).into();
```
it doesn't seem ideal that this is hard-coded. somebody will want to customize it.

tingerrr replied:
Would using `Finalize` and setting it there be appropriate? That should enable `show quote: set pad(x: ...)` to change it, if I understood the usage in `figure.rs` right.

laurmaedje replied:
Finalize is a bit wonky, but it would probably work. Should be documented though then.

laurmaedje commented on crates/typst-library/src/text/quote.rs:
```diff
@@ -0,0 +1,136 @@
+                ])
+            };
+
+            if TextElem::dir_in(styles) == Dir::LTR {
+                realized = inline(realized, author);
+            } else {
+                realized = inline(author, realized);
+            }
```
Do we really need explicit code for the swapping? Without it, an English quote would be LTR in an RTL context, but an Arabic quote wouldn't be. The swapping already happens implicitly because the content ends up right to left. It just isn't obversable for English text because BiDi is smart enough not to write English in RTL.

tingerrr replied:
I'm not quite sure when exactly content is rearranged for RTL, I noticed that in RTL:
- `[a --- b]` stays `[a --- b]`
- `[a ---]` becomes `[--- a]`

I asked a colleague of mine from Jordan if inline quotes would follow my intuition of being `[author --- "quote"]` in say arabic.

So I want to display `[#quote --- #author]` for LTR and `[#author --- #quote]` for RTL, this doesn't happen automatically. For the block version, no swapping is needed because the em dash is leading.

But maybe it can be done better, I'm up for suggestions.

laurmaedje replied:
`a --- b` stays `a --- b` because `a` and `b` are Latin letters and Latin is not an RTL script. Setting RTL doesn't force Latin to be written from right-to-left, it only sets the top-level text direction (which affects punctuation), but embedded LTR content stays RTL. If you try the same thing with actual RTL text, it is swapped:

 ```typst
#set text(lang: "ar", font: "Noto Sans Arabic")

// Each word in its own line so that the order is clear.
مرحبًا 
---
عالم

```
Result:
<img width="150" alt="image" src="https://github.com/typst/typst/assets/17899797/da474eba-34f8-4f45-b870-3c1ea4d5b21f">

reknih commented:
I had a look at the styles and the arguments and have a few remarks:

- The author attribution in the inline quotation looks awkward. Maybe it should just be suppressed.
- The block-level quote is missing spacing above. It is too tight with the surrounding paragraphs atm. You may need to use disgression to insert spacing below as well.
- The author attribution for block-level quotes should be closer to the quote.
- Inline quotes suppress their source if they are missing an author.
- The source is appended directly after the text. It should be separated from the quote by a space, and maybe a comma.
- For block-level quotes, the source should be placed in the author attribution.

I'm not sure whether the combination of author and source attributes is the best solution for attribution. First of all, the collision with URLs in source can be solved by passing a label:

```typst
#quote(block: true, author: "Honoré de Balzac", source: <humaine>)[Bonjour]

or maybe even

#quote(block: true, author: "Honoré de Balzac")[Bonjour] @humaine
```

The latter alternative uses the fact that the reference syntax already semantically attaches the reference to the content before it. The quote should be rendered like this:

```
   Bonjour
          -- Honoré de Balzac, [1]
```

However, we will run into problems with citation styles like `chicago-author-date` which a) already contain the author and b) are parenthesized which results in a bit of a mess. CSL (implementation upcoming) is already able to differentiate these styles with the [`citation-format` attribute](https://docs.citationstyles.org/en/stable/specification.html#info), for the current styles we'd need to do it manually. Given the styles there, I propose the following behavior:

- `numeric` and `label`: Put the author, then a comma, and then the unchanged citation (e.g. "belle hooks, [1]")
- `note`: Put the author and then the footnote (e.g. "Thomas Mann¹")
- `author-date`, `author`: Suppress the author, remove any parentheses around the cite and print it ("Einstein and Laub, 1908")

Arguabely, the author argument is not needed at all if we got a reference since we can extract the author from the Hayagriva database, but I'll leave this up to discussion.

tingerrr commented:
> The author attribution in the inline quotation looks awkward. Maybe it should just be suppressed.

Fine with me, the user can just use a show rule after all.

> The block-level quote is missing spacing above. It is too tight with the surrounding paragraphs atm. You may need to use disgression to insert spacing below as well.

I went with `0.5em` padding here, you'll likely have to tell me the appropriate defaults for this, as I don't really have an eye for spacing like that. I'm not sure what is meant with digression in a typesetting context.

> The author attribution for block-level quotes should be closer to the quote.

Once again, I went with `h(1em, weak: true)`, this looks at least to my eye like the regular line height. But I think this is font dependent, right? Regardless, I need some advice what you guys think is right here.

> Inline quotes suppress their source if they are missing an author.

Yup, that should be fixed now.

> For block-level quotes, the source should be placed in the author attribution.

That was actually intended, not sure how I overlooked that while testing, I added more test cases.

So I went with the label solution, I forgot that I could simply extract the string from it to use in `CiteElem::new`. The problem with the `@label` solution is that I don't know whether the next element is a `CiteElem`. Since the quote is wrapped in `PadElem` it would not display like that without knowing this and "pulling it inside" so to speak.

tingerrr commented:
> I'm not sure whether the combination of author and source attributes is the best solution for attribution.

I think you're right, it probably makes sense to merge `author` and `source` into an `attribution` property. On Wikipedia for "I know that I know nothing" we see this:

> ... ἔοικα γοῦν τούτου γε σμικρῷ τινι αὐτῷ τούτῳ σοφώτερος εἶναι, ὅτι ἃ μὴ οἶδα οὐδὲ οἴομαι εἰδέναι.
... I seem, then, in just this little thing to be wiser than this man at any rate, that what I do not know I do not think I know either. [from the Henry Cary literal translation of 1897]

The attribution is not just the author but additional information about the translation, so just `author` for the field makes less sense.

Attributions could work like this:
- `attribution: <humaine>`, equivalent to `attribution: {set ...; [@humaine]}` with the specified `set cite(bracket: false)` for some styles
- `attribution: [...]`, anything else

If the user wants to link the whole quote, they can do it themselves instead. For inline quotes, a bib attribution would simply append the attribution as is. But it should probably not be suppressed by default the way author was.

One problem, I see with this, is that `attribution: <humaine>` for non-author styles would render just `-- [1]` or `--¹`. We could simply warn on those styles when only a label is passed.

Another problem is that a change in citation style between those would now need manual edits to all attributions. Perhaps the attibution should have an optional prefix which is only used for the non-author styles:

- `quote(attribution: <humaine>)` equaivalent to `quote(attribution: (prefix: none, label: <humaine>)`
- `quote(attribution: (prefix: [Honoré de Balzac], label: <humaine>))`

resulting in `Honoré de Balzac¹`/ `Honoré de Balzac, [1]` on notes/numeric and in `Honoré de Balzac`/`Honoré de Balzac, #date` on author/author-date.

But this seems like quite a lot of rules to keep in mind as a user and I'm wondering if you have better ideas.

reknih commented:
> I went with `0.5em` padding here, you'll likely have to tell me the appropriate defaults for this, as I don't really have an eye for spacing like that. I'm not sure what is meant with digression in a typesetting context.

I pushed some spacing constants. By disgression I just meant that you'd need to check if the below spacing still looked good with another above spacing.

> Attributions could work like this:

Sounds good

> One problem, I see with this, is that attribution: `<humaine>` for non-author styles would render just -- `[1]` or --`¹`. We could simply warn on those styles when only a label is passed.

We can do that for the moment, but we'd want to retrieve the entry and print the authors in the future.

> Another problem is that a change in citation style between those would now need manual edits to all attributions. Perhaps the attribution should have an optional prefix which is only used for the non-author styles

Given the future behavior, I think we can do without a prefix right now.

tingerrr commented:
> By disgression I just meant that you'd need to check if the below spacing still looked good with another above spacing.

I added blocks with no spacing (left) and `1em` spacing (right) around the `lorem`, the spacing seems fine to me, but since the defaults are very small it is hard to see, once we go above those they collapse the padding it seems, I'm not fully aware of how weakness works for vertical spacing.
![image](https://github.com/typst/typst/assets/137803093/2f783ec6-69fc-41ee-aac4-eb57984878c0)

> We can do that for the moment, but we'd want to retrieve the Hayagriva entry and print the authors in the future.

I left a `TODO` in the section that defines the style, linking to the comments above. I've added a small example for and note explaining that `-- [1]` and `--¹` will be fixed and how they can be worked around for now.

reknih reviewed (approved):
Looks good now!

laurmaedje commented on crates/typst-library/src/text/quote.rs:
```diff
@@ -0,0 +1,209 @@
+                                BibliographyElem::find(vt.introspector).at(self.span())?;
+
+                            // TODO: these should use the citation-format attribute, once CSL
+                            // is implemented and retreive the authors for non-author formats
```
```suggestion
                            // is implemented and retrieve the authors for non-author formats
```

laurmaedje commented on crates/typst-library/src/text/quote.rs:
```diff
@@ -0,0 +1,209 @@
+                }
+
+                // use v(0.9em, weak: true) bring the attribution closer to the quote
+                let weak_v = VElem::weak(Spacing::Rel(Em::new(0.9).into())).pack();
```
Is this value customizable without redefining the whole quote?

tingerrr replied:
No, it is not, but I hardly see someone adjusting only this and not the whole quote. `0.9` was chosen by @reknih. Adding a parameter for just that doesn't sound right, it's very much tied to this specific default show implementation.

laurmaedje replied:
No, a parameter for this does indeed not sound right. Just wondering. I guess it's fine. I have just noticed how as soon as there's any hardcoded number or separator in a default show rule, somebody opens an issue about customization.

laurmaedje reviewed (commented):
Labels definitely make sense for attribution. I wonder whether `cite` should also take labels instead of strings ...

tingerrr commented:
> Labels definitely make sense for attribution. I wonder whether `cite` should also take labels instead of strings ...

Yes, that makes sense, citing keys with spaces is still possible although more cumbersome (`cite(label("key with spaces"))`), but that should be fine no one uses spaces in their bib keys riiiight?

laurmaedje commented:
Thank you!
